google/cloud/ndb/_transaction.py

```diff
@@ -23,6 +23,123 @@
+
+        propagation_options = context_module.TransactionOptions._PROPAGATION
+        if propagation is None or propagation in propagation_options:
+            self.propagation = propagation
+        else:
+            raise ValueError(
+                "Unexpected value for propagation. Got: {}. Expected one of: "
+                "{}".format(propagation, propagation_options)
+            )
+
+        propagation_names = context_module.TransactionOptions._INT_TO_NAME
+        self.propagation_name = propagation_names.get(self.propagation)
+
+        self.join = join
+        joinable_options = context_module.TransactionOptions._JOINABLE
+        self.joinable = propagation in joinable_options
+        super(_Propagation, self).__init__(*args, **kwargs)
+
+    def _handle_nested(self):
+        """The NESTED propagation policy would commit all changes in the outer
+        and inner transactions together when the outer policy commits. However,
+        if an exception is thrown in the inner transaction all changes there
+        would get thrown out but allow the outer transaction to optionally
+        recover and continue. The NESTED policy is not supported. If you use
+        this policy, your code will throw a BadRequestError exception.
+        """
+        raise exceptions.BadRequestError("Nested transactions are not supported.")
+
+    def _handle_mandatory(self):
+        """Always propagate an existing transaction; throw an exception if
+        there is no existing transaction. If a function that uses this policy
+        throws an exception, it's probably not safe to catch the exception and
+        commit the outer transaction; the function may have left the outer
+        transaction in a bad state.
+        """
+        if not in_transaction():
+            raise exceptions.BadRequestError("Requires an existing transaction.")
+
+    def _handle_allowed(self):
+        """If there is an existing transaction, propagate it. If a function
+        that uses this policy throws an exception, it's probably not safe to
+        catch the exception and commit the outer transaction; the function may
+        have left the outer transaction in a bad state.
+        """
+        # no special handling needed.
+        pass
+
+    def _handle_independent(self):
+        """Always use a new transaction, "pausing" any existing transactions.
+        A function that uses this policy should not return any entities read in
+        the new transaction, as the entities are not transactionally consistent
+        with the caller's transaction.
+        """
+        if in_transaction():
+            # Avoid circular import in Python 2.7
+            from google.cloud.ndb import context as context_module
+
+            ctx = context_module.get_context()
+            ctx.flush()
+            new_ctx = ctx.new(
+                transaction=None, batches=None, commit_batches=None, cache=None
+            )
```

Comment:
@cguardia / @andrewsg this small block is the part I am most nervous about / not sure if I got right.

My concerns are:

1. The comment for [`non_transactional`](https://github.com/googleapis/python-ndb/blob/448054faf8da812a387614681d0771dcbfa18ef2/google/cloud/ndb/_transaction.py#L310-L311) says that it pauses the existing transaction but looking at the implementation it doesn't actually call `flush` on the previous context. Whereas the old ndb library - who's comment also says it pauses the transaction - does [call flush](https://github.com/GoogleCloudPlatform/datastore-ndb-python/blob/05c0f9ebd1fcd746367b79d8b7396f6d51e22188/ndb/context.py#L826-L841). **Am I pausing the previous transaction correctly?**

2. [`non_transactional`](https://github.com/googleapis/python-ndb/blob/448054faf8da812a387614681d0771dcbfa18ef2/google/cloud/ndb/_transaction.py#L333) only sets `transaction=None`, but I wasn't sure if that was enough or not to make the transactions truly independent / mimic the old behaviour, particularly in regard to shared cache. So as a precaution - _rather than because I knew what I was doing_ - I also set the parameters `batches`, `commit_batches` and `cache` to `None`. **Should I have left these parameters to just take their defaults?**

When you complete your review(s) please allay my concerns?

Reply:
@chrisrossi can you give us your opinion on these questions? This is almost ready to merge.

Reply:
Regarding 1), I don't believe the `flush` is necessary. I don't think anything would break either way, but by not calling `flush`, we let the parent context continue lazily add to its batches after the sub-transaction (or non-transaction) has finished executing, and we don't pause to wait for Datastore before starting the sub-transaction (or non-transaction).

Reply:
> I don't believe the flush is necessary

Ok, I think I agree now too, I will go with the same behaviour as in the new [`non_transactional`](https://github.com/googleapis/python-ndb/blob/448054faf8da812a387614681d0771dcbfa18ef2/google/cloud/ndb/_transaction.py#L310-L311) implementation (no flush)

> I think you should be able to just set transaction ... It's worth double checking my logic on that one.

Yep, checked. You were right.